motor/render/vulkan_renderer.cpp

```diff
@@ -36,6 +37,96 @@ struct DepthBuffer
   vk::ImageView view_;
 };
 
+struct QueueFamilyIndices
+{
+  std::optional<size_t> graphics_queue_idx;
+  std::optional<size_t> present_queue_idx;
+
+  bool isComplete()
+  {
+    return graphics_queue_idx.has_value() && present_queue_idx.has_value();
+  }
+};
+
+std::vector<const char*> GetRequiredExtensions()
+{
+  std::vector<const char*> extensions;
+  uint32_t glfw_extension_count = 0;
+  const char** glfw_extensions =
+      glfwGetRequiredInstanceExtensions(&glfw_extension_count);
+  CHECK(glfw_extensions) << "Vulkan is not supported on the machine";
+  extensions.resize(glfw_extension_count);
+  std::copy(glfw_extensions, glfw_extensions + glfw_extension_count,
+            extensions.begin());
+
+  return extensions;
+}
+
+QueueFamilyIndices GetQueueIndices(const vk::PhysicalDevice& phy_dev,
+                                   const vk::SurfaceKHR& surface)
+{
+  QueueFamilyIndices indices;
+  std::vector<vk::QueueFamilyProperties> queue_family_props =
+      phy_dev.getQueueFamilyProperties();
+  CHECK(!queue_family_props.empty()) << "No queues on the device";
+
+  for (size_t i = 0; i < queue_family_props.size(); ++i)
+  {
+    const auto& family_prop = queue_family_props[i];
+    if (family_prop.queueCount &&
+        family_prop.queueFlags & vk::QueueFlagBits::eGraphics)
+    {
+      indices.graphics_queue_idx = i;
+    }
+    if (VkSuccuessOrDie(phy_dev.getSurfaceSupportKHR(i, surface),
+                        "Couldn't query for KHR support"))
+    {
+      indices.present_queue_idx = i;
+    }
+  }
+  return indices;
+}
+
+bool CheckExtensionSupport(const vk::PhysicalDevice& phy_dev)
+{
+  std::vector<vk::ExtensionProperties> available_extensions =
+      VkSuccuessOrDie(phy_dev.enumerateDeviceExtensionProperties(),
+                      "Coudn't get device extensions");
+  // device_extensions is also used when creating logical device, maybe move it
```

Comment:
yeah feel free to move that into a global `constexpr const char* extensions[] = { ... }`.